lib/spack/spack/database.py

```diff
@@ -486,8 +497,19 @@ def check(cond, msg):
         if version > _db_version:
             raise InvalidDatabaseVersionError(_db_version, version)
         elif version < _db_version:
-            self.reindex(spack.store.layout)
-            installs = dict((k, v.to_dict()) for k, v in self._data.items())
+            if not any(
+                    old == version and new == _db_version
+                    for old, new in _skip_reindex
+            ):
+                tty.warn(
```

Comment:
Tried this out and it works correctly. The only minor user facing thing that can be improved imho is:
```suggestion
                warnings.warn(
```
to avoid the warning being printed twice (I tweaked the version numbers to get the print below):
```
$ spack find
==> Warning: Spack database version changed from 0.9.3 to 5. Upgrading.
==> Warning: Spack database version changed from 0.9.3 to 5. Upgrading.
==> 3 installed packages
-- linux-ubuntu18.04-broadwell / gcc@9.0.1 ----------------------
hdf5@1.10.5  libszip@2.1.1  zlib@1.2.11
```
This seem to happen when a read transaction needs to `reindex` and triggers a nested write transaction. We can probably simplify the logic at a later point if need be.